client/modules/IDE/components/Editor/index.jsx

```diff
@@ -81,36 +76,128 @@ window.HTMLHint = HTMLHint;
+  const editorAccessibility = useSelector((state) => state.editorAccessibility);
+  const sidebarIsExpanded = useSelector((state) => state.ide.sidebarIsExpanded);
+  const consoleEvents = useSelector((state) => state.console);
+  const mergedState = {
+    ...preferences,
+    ...ide,
+    ...project,
+    ...editorAccessibility,
+    isExpanded: sidebarIsExpanded
+  };
+  // Binding Action Creators
+  const funcprop = bindActionCreators(
+    Object.assign(
+      {},
+      EditorAccessibilityActions,
+      FileActions,
+      ProjectActions,
+      IDEActions,
+      PreferencesActions,
+      UserActions,
+      ConsoleActions
+    ),
+    dispatch
+  );
+  // refs for codemirror
+  const cm = useRef(null);
+  // defining functions
+  function getFileMode(fileName) {
+    let mode;
+    if (fileName.match(/.+\.js$/i)) {
+      mode = 'javascript';
+    } else if (fileName.match(/.+\.css$/i)) {
+      mode = 'css';
+    } else if (fileName.match(/.+\.(html|xml)$/i)) {
+      mode = 'htmlmixed';
+    } else if (fileName.match(/.+\.json$/i)) {
+      mode = 'application/json';
+    } else if (fileName.match(/.+\.(frag|glsl)$/i)) {
+      mode = 'x-shader/x-fragment';
+    } else if (fileName.match(/.+\.(vert|stl)$/i)) {
+      mode = 'x-shader/x-vertex';
+    } else {
+      mode = 'text/plain';
+    }
+    return mode;
+  }
+  // States and refs
+  const codemirrorContainer = useRef(null);
 
-    this.updateLintingMessageAccessibility = debounce((annotations) => {
-      this.props.clearLintMessage();
-      annotations.forEach((x) => {
-        if (x.from.line > -1) {
-          this.props.updateLintMessage(x.severity, x.from.line + 1, x.message);
+  function prettierFormatWithCursor(parser, plugins) {
+    try {
+      const { formatted, cursorOffset } = prettier.formatWithCursor(
+        cm.current.doc.getValue(),
+        {
+          cursorOffset: cm.current.doc.indexFromPos(cm.current.doc.getCursor()),
+          parser,
+          plugins
         }
-      });
-      if (this.props.lintMessages.length > 0 && this.props.lintWarning) {
-        this.beep.play();
-      }
-    }, 2000);
-    this.showFind = this.showFind.bind(this);
-    this.showReplace = this.showReplace.bind(this);
-    this.getContent = this.getContent.bind(this);
+      );
+      const { left, top } = cm.current.getScrollInfo();
+      cm.current.doc.setValue(formatted);
+      cm.current.focus();
+      cm.current.doc.setCursor(cm.current.doc.posFromIndex(cursorOffset));
+      cm.current.scrollTo(left, top);
+    } catch (error) {
+      console.error(error);
+    }
+  }
+  function tidyCode() {
+    const mode = cm.current.getOption('mode');
+    if (mode === 'javascript') {
+      prettierFormatWithCursor('babel', [babelParser]);
+    } else if (mode === 'css') {
+      prettierFormatWithCursor('css', [cssParser]);
+    } else if (mode === 'htmlmixed') {
+      prettierFormatWithCursor('html', [htmlParser]);
+    }
+  }
```

Comment:
This could potentially be moved outside of the component as well, it you were to take the CodeMirror instance as an argument.